src/tests/conformance/handshake/ignore_message_inplace_of_verack.rs

```diff
@@ -0,0 +1,300 @@
+        let block_hash = Block::testnet_genesis().double_sha256().unwrap();
+        let block_loc = LocatorHashes::new(vec![block_hash], Hash::zeroed());
+        run_test_case(Message::GetHeaders(block_loc)).await;
+    }
+
+    #[tokio::test]
+    async fn get_blocks() {
+        // zcashd: pass
+        // zebra:  fail (disconnects)
+        let block_hash = Block::testnet_genesis().double_sha256().unwrap();
+        let block_loc = LocatorHashes::new(vec![block_hash], Hash::zeroed());
+        run_test_case(Message::GetBlocks(block_loc)).await;
+    }
+
+    #[tokio::test]
+    async fn get_data_block() {
+        // zcashd: pass
+        // zebra:  fail (disconnects)
+        let block_inv = Inv::new(vec![Block::testnet_genesis().inv_hash()]);
+        run_test_case(Message::GetData(block_inv)).await;
+    }
+
+    #[tokio::test]
+    async fn get_data_tx() {
+        // zcashd: pass
+        // zebra:  fail (disconnects)
+        run_test_case(Message::GetData(Inv::new(vec![Block::testnet_genesis()
+            .txs[0]
+            .inv_hash()])))
+        .await;
+    }
+
+    #[tokio::test]
+    async fn inv() {
+        // zcashd: pass
+        // zebra:  fail (disconnects)
+        let block_inv = Inv::new(vec![Block::testnet_genesis().inv_hash()]);
+        run_test_case(Message::Inv(block_inv)).await;
+    }
+
+    #[tokio::test]
+    async fn not_found() {
+        // zcashd: pass
+        // zebra:  fail (disconnects)
+        let block_inv = Inv::new(vec![Block::testnet_genesis().inv_hash()]);
+        run_test_case(Message::NotFound(block_inv)).await;
+    }
+
+    /// Checks that `message` gets ignored when sent instead of [`Message::Version`] when the node
+    /// receives the connection.
+    async fn run_test_case(message: Message) {
+        // Spin up a node instance.
+        let mut node = Node::new().unwrap();
+        node.initial_action(Action::WaitForConnection)
+            .start()
+            .await
+            .unwrap();
+        // Connect to the node, and exchange versions.
+        let mut synthetic_node = SyntheticNode::builder()
+            .with_version_exchange_handshake()
+            .build()
+            .await
+            .unwrap();
+        synthetic_node.connect(node.addr()).await.unwrap();
+
+        // Send a non-verack message.
+        synthetic_node
+            .send_direct_message(node.addr(), message)
+            .await
+            .expect("Sending non-version message");
+
+        // Expect the node to ignore the previous message, verify by completing the handshake.
+        // Send Verack.
+        synthetic_node
+            .send_direct_message(node.addr(), Message::Verack)
+            .await
+            .expect("Sending Verack");
+
+        // Read Verack.
+        let (_, verack) = synthetic_node
+            .recv_message_timeout(RECV_TIMEOUT)
+            .await
+            .expect("Receiving Verack");
+        assert_matches!(verack, Message::Verack);
```

Comment:
In line with the comment above, we should probably return an error instead of panicking?